loki/transform/transform_hoist_variables.py

```diff
@@ -0,0 +1,335 @@
+                item.user_data[self._key]["hoist_variables"]))
+
+    def find_variables(self, routine):
+        """
+        **Override**: Find/Select all the variables to be hoisted.
+
+        Selects all local variables that are not ``parameter`` to be hoisted.
+
+        Parameters
+        ----------
+        routine : :any:`Subroutine`
+            The subroutine find the variables.
+        """
+        return [var for var in routine.variables if var not in routine.arguments if not var.type.parameter]
+
+
+class HoistVariablesTransformation(Transformation):
+    """
+    **Base class** for the *Synthesis* part of the hoist variables functionality/transformation.
+
+    Traverses all subroutines to hoist the variables.
+    Create a derived class and override :func:`find_variables<HoistVariablesAnalysis.find_variables>`
+    to define which variables to be hoisted.
+
+    .. note::
+        Needs the *Analysis* part to be processed first in order to hoist all already found variables.
+
+    Parameters
+    ----------
+    key : str
+        Access identifier/key for the ``item.user_data`` dictionary. Only necessary to provide if several of
+        these transformations are carried out in succession.
+    """
+
+    _key = 'HoistVariablesTransformation'
+
+    def __init__(self, key=None):
+        if key is not None:
+            self._key = key
+
+    def transform_subroutine(self, routine, **kwargs):
+        """
+        Transformation applied to :any:`Subroutine` item.
+
+        Hoists all to be hoisted variables which includes
+
+        * appending the arguments for each subroutine
+        * appending the arguments for each subroutine call
+        * modifying the variable declaration in the subroutine
+        * adding the variable declaration in the driver
+
+        Parameters
+        ----------
+        routine : :any:`Subroutine`
+            The subroutine to be transformed.
+        **kwargs : optional
+            Keyword arguments for the transformation.
+        """
+        role = kwargs.get('role', None)
+        item = kwargs.get('item', None)
+        successors = kwargs.get('successors', ())
+        successor_map = {successor.routine.name: successor for successor in successors}
+
+        if item and not item.local_name == routine.name.lower():
+            return
+
+        if self._key not in item.user_data:
+            raise RuntimeError(f'{self.__class__.__name__} requires key "{self._key}" in item.user_data!\n'
+                               f'Make sure to call HoistVariablesAnalysis (or any derived class) before and to provide '
+                               f'the correct key.')
+
+        if role == 'driver':
+            for var in item.user_data[self._key]["to_hoist"]:
+                self.driver_variable_declaration(routine, var)
+        else:
+            routine.arguments += as_tuple([var.clone(type=var.type.clone(intent='inout', allocatable=None),
```

Comment:
Do we need this to be an `allocatable` here? Variables with an `intent` attribute can be declared without, I think.